### PR TITLE
[4.0][webservices] allow public GET with rate limit control

### PR DESCRIPTION
**Pull request by alikon**
Pull Request for Issue #27569  point 3).

### Summary of Changes

-  endpoint fine grained control
-  allow public GET for an endpoint
-  endpoint rate limit control
- CLI command to reset limit quota `php cli/joomla.php webservices:reset`

enabled for:
 

- `plg_webservices_content`
- `plg_webservices_contact`
- `plg_webservices_modules`
- `plg_webservices_users`

the others webservices plugin works as before (ie no public GET)

### Testing Instructions

- apply PR or download the Prebuilt package
- generate a new Joomla API TOKEN
- go to plugins - Web Services - Content/Contact/Modules/Users
- check the new config parameters

image

### Endpoint Allowed Verbs Test
- select all verbs and check that each one works as before
- check that only the allowed verbs works
-- enable POST only
-- and make a GET {{base_path}}/api/index.php/v1/content/article 
-- you'll get a **404 - Resource not found**

### Public GET Test
- enable Allow Public GET
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **without** the Joomla API TOKEN (noAuth)
-- you'll get a **200 response**
-- check that only public content is returned

- enable Allow Public GET
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **with** the Joomla API TOKEN 
-- you'll get a **200 response**
-- check that all content is returned (punblishe/unpublished)

- disable Allow Public GET
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **without** the Joomla API TOKEN (noAuth)
-- you'll get a **401 response**

- disable Allow Public GET
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **with** the Joomla API TOKEN 
-- you'll get a **200 response**

### Rate Limit Test
- enable Allow Public GET
- set Rate limit to 5
-- and make 6 GET {{base_path}}/api/index.php/v1/content/article  **without** the Joomla API TOKEN (noAuth)
-- you'll get a **429 response**
- click on Reset Limit
-- and make a GET {{base_path}}/api/index.php/v1/content/article  **without** the Joomla API TOKEN (noAuth)
-- you'll get a **200 response**

### Actual result BEFORE applying this Pull Request

n/a

### Expected result AFTER applying this Pull Request

you can allow public GET
you can choose what verbs can be requested for each plugin

### Documentation Changes Required
yes

### CODE REVIEW

**Comment by richard67** [comment]
@alikon PHPCS https://ci.joomla.org/joomla/joomla-cms/38066/1/6

File: administrator/language/en-GB/plg_webservices_content.ini [hunk 1]
```diff
@@ -5,3 +5,10 @@
 
 PLG_WEBSERVICES_CONTENT="Web Services - Content"
 PLG_WEBSERVICES_CONTENT_XML_DESCRIPTION="Add articles routes to the API for your website."
+PLG_WEBSERVICES_ALLOWEDVERBS_LABEL = "Allowed verbs"
```
Comment (richard67) [inline administrator/language/en-GB/plg_webservices_content.ini]:
```suggestion
PLG_WEBSERVICES_ALLOWEDVERBS_LABEL = "Allowed Verbs"
```
Labels should be capitalized.

File: administrator/language/en-GB/plg_webservices_content.ini [hunk 2]
```diff
@@ -5,3 +5,10 @@
...
+PLG_WEBSERVICES_FIELD_VALUE_GET = "GET"
+PLG_WEBSERVICES_FIELD_VALUE_POST = "POST"
+PLG_WEBSERVICES_FIELD_VALUE_DELETE = "DELETE"
+PLG_WEBSERVICES_FIELD_VALUE_PUT = "PUT"
+PLG_WEBSERVICES_FIELD_VALUE_PATCH = "PATCH"
+PLG_WEBSERVICES_PUBLIC_LABEL = "Allow public GET"
```
Comment (richard67) [inline administrator/language/en-GB/plg_webservices_content.ini]:
```suggestion
PLG_WEBSERVICES_PUBLIC_LABEL = "Allow Public GET"
```
Labels should be capitalized.

File: libraries/src/Router/ApiRouter.php [hunk 3]
```diff
@@ -58,17 +58,32 @@ public function __construct(CMSApplicationInterface $app, array $maps = [])
 	 *
 	 * @since   4.0.0
 	 */
-	public function createCRUDRoutes($baseName, $controller, $defaults = [], $publicGets = false)
+	public function createCRUDRoutes($baseName, $controller, $defaults = [], $publicGets = false, $allowed = [])
```
Comment (richard67) [inline libraries/src/Router/ApiRouter.php:62]: The new `$allowed` parameter should be added to the doc block of this function.

File: plugins/webservices/content/content.php [hunk 4]
```diff
@@ -37,18 +69,27 @@ class PlgWebservicesContent extends CMSPlugin
 	 *
 	 * @since   4.0.0
 	 */
-	public function onBeforeApiRoute(&$router)
+	public function onBeforeApiRoute(&$router, $object)
```
Comment (richard67) [inline plugins/webservices/content/content.php]: The new `$object` parameter should be added to the doc block of this function.

File: libraries/src/Router/ApiRouter.php (hunk 3 above)
Comment (alikon) [reply libraries/src/Router/ApiRouter.php:62]: done

Comment (alikon) [reply libraries/src/Router/ApiRouter.php:62]: done

File: administrator/language/en-GB/plg_webservices_content.ini [hunk 5]
```diff
@@ -5,3 +5,10 @@
 
 PLG_WEBSERVICES_CONTENT="Web Services - Content"
 PLG_WEBSERVICES_CONTENT_XML_DESCRIPTION="Add articles routes to the API for your website."
+PLG_WEBSERVICES_ALLOWEDVERBS_LABEL = "Allowed Verbs"
+PLG_WEBSERVICES_FIELD_VALUE_GET = "GET"
+PLG_WEBSERVICES_FIELD_VALUE_POST = "POST"
+PLG_WEBSERVICES_FIELD_VALUE_DELETE = "DELETE"
+PLG_WEBSERVICES_FIELD_VALUE_PUT = "PUT"
+PLG_WEBSERVICES_FIELD_VALUE_PATCH = "PATCH"
```
Comment (richard67) [inline administrator/language/en-GB/plg_webservices_content.ini]: As "GET", "POST", "DELETE", "PUT" and "PATCH" are fixed terms which should not be translated: Wouldn't it be better to just use these terms hard-coded in the XML and not use language strings? Similar use case are the robots options, see #30110 et al.

@brianteeman Your opinion?

File: libraries/src/Router/ApiRouter.php [hunk 6]
```diff
@@ -53,22 +53,38 @@ public function __construct(CMSApplicationInterface $app, array $maps = [])
 	 * @param   string  $controller  The name of the controller that contains CRUD functions.
 	 * @param   array   $defaults    An array of default values that are used when the URL is matched.
 	 * @param   bool    $publicGets  Allow the public to make GET requests.
+	 * @param   array   $defaults    An array of allowed verbs.
```
Comment (richard67) [inline libraries/src/Router/ApiRouter.php]:
```suggestion
	 * @param   array   $allowed     An array of allowed verbs.
```

File: administrator/language/en-GB/plg_webservices_content.ini (hunk 5 above)
Comment (alikon) [reply administrator/language/en-GB/plg_webservices_content.ini]: this can be done later, if and when this RFC get some positive feedback from RL

File: plugins/webservices/content/content.php (hunk 4 above)
Comment (alikon) [reply plugins/webservices/content/content.php]: done

**Comment by richard67** [comment]
To me it looks good, and I like the approach. I'm not deep enough into API and security stuff to say if it's safe enough or not, but I can't see any issue with it right now.

**Comment by richard67** [comment]
API tests are failing https://ci.joomla.org/joomla/joomla-cms/38077/1/25 ... I guess this is expected due to tests having to be adjusted to the changes in this PR?

**Comment by wilsonge** [comment]
Sorry I've made my point badly there. If you set the public param (at least back then) the endpoint couldn't be accessed. I don't think this should be enabled for any core endpoints. For example with this change you'd allow public get's of unpublished articles (I assume - haven't tested)

**Comment by alikon** [comment]
not sure i fully get you 

with the public GET allowed you can simply use `{{base_path}}/api/index.php/v1/content/article?filter[state]=1`

File: administrator/language/en-GB/plg_webservices_content.ini (hunk 5 above)
Comment (alikon) [reply administrator/language/en-GB/plg_webservices_content.ini]: done

File: administrator/language/en-GB/plg_webservices_content.ini [hunk 7]
```diff
@@ -5,3 +5,5 @@
 
 PLG_WEBSERVICES_CONTENT="Web Services - Content"
 PLG_WEBSERVICES_CONTENT_XML_DESCRIPTION="Add articles routes to the API for your website."
+PLG_WEBSERVICES_CONTENT_ALLOWEDVERBS_LABEL = "Allowed Verbs"
+PLG_WEBSERVICES_CONTENT_PUBLIC_LABEL = "Allow Public GET"
```
Comment (Quy) [inline administrator/language/en-GB/plg_webservices_content.ini]:
```suggestion
PLG_WEBSERVICES_CONTENT_ALLOWEDVERBS_LABEL="Allowed Verbs"
PLG_WEBSERVICES_CONTENT_PUBLIC_LABEL="Allow Public GET"
```

**Comment by snehal0904** [comment]
I have tested this item :white_check_mark: successfully on 1edf1fc4898fcaaabee6b828aedc28eb6dea7673<br /><br />Working correctly. But not sure about security.

Thank you for the implementation<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31579">issues.joomla.org/tracker/joomla-cms/31579</a>.</sub>

**Comment by shinderohanoffical** [comment]
I have tested this item :white_check_mark: successfully on 1edf1fc4898fcaaabee6b828aedc28eb6dea7673<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31579">issues.joomla.org/tracker/joomla-cms/31579</a>.</sub>

**Comment by richard67** [comment]
> not sure i fully get you
> 
> with the public GET allowed you can simply use `{{base_path}}/api/index.php/v1/content/article?filter[state]=1`

@alikon I think the question George had was if unpublished articles will be shown to public if not using such a filter. This should not be the case, i.e. public never should see unpublished stuff.

**Comment by alikon** [comment]
ok unpublished content are not shown when public GET

File: libraries/src/Application/ApiApplication.php [hunk 8]
```diff
@@ -292,15 +292,21 @@ protected function route()
...
 		{
-			if (!$this->login(array('username' => ''), array('silent' => true, 'action' => 'core.login.api')))
+			if (!isset($route['vars']['public']) || $route['vars']['public'] === false)
 			{
 				throw new AuthenticationFailed;
-			}
+			}			
+			if ($route['vars']['public'] === true)
+			{
+				$publicApi = false;
```
Comment (wilsonge) [inline libraries/src/Application/ApiApplication.php]: Isn't this the wrong way around? `$publicApi` should be true if we end up here? Also shouldn't this be an else rather than a separate if?

File: libraries/src/Router/ApiRouter.php [hunk 9]
```diff
@@ -53,22 +53,38 @@ public function __construct(CMSApplicationInterface $app, array $maps = [])
...
 	 * @param   array   $defaults    An array of default values that are used when the URL is matched.
 	 * @param   bool    $publicGets  Allow the public to make GET requests.
+	 * @param   array   $allowed     An array of allowed verbs.
 	 *
 	 * @return  void
 	 *
 	 * @since   4.0.0
 	 */
-	public function createCRUDRoutes($baseName, $controller, $defaults = [], $publicGets = false)
+	public function createCRUDRoutes($baseName, $controller, $defaults = [], $publicGets = false, $allowed = [])
```
Comment (wilsonge) [inline libraries/src/Router/ApiRouter.php:62]: What's the justification behind this allowedRoutes stuff? Like I'm kinda unconvinced as to what it's use cases are?

Comment (alikon) [reply libraries/src/Router/ApiRouter.php:62]: The use case for me is:
to allow more fine grained control on what are the verbs that i want to expose for that endpoint.

As an example, suppose that:

for the endpoint A)
---- i ONLY want permit GET and not POST, PATCH,DELETE etc
but 
for the endpoint B) 
---- i want permit GET and POST only
and/or
for the endpoint C)
---- i want permit GET, POST and PATCH but not DELETE
etc...

as it is now you always have the same configuration for all the endpoints enabled

File: libraries/src/Application/ApiApplication.php (hunk 8 above)
Comment (alikon) [reply libraries/src/Application/ApiApplication.php]: i'm going to update this pr _"quite a bit"_ for allowing a rate limit control for the pubic GET for example
let's rediscuss this after that

File: administrator/language/en-GB/plg_webservices_content.ini [hunk 10]
```diff
@@ -5,3 +5,6 @@
 
 PLG_WEBSERVICES_CONTENT="Web Services - Content"
 PLG_WEBSERVICES_CONTENT_XML_DESCRIPTION="Add articles routes to the API for your website."
+PLG_WEBSERVICES_CONTENT_ALLOWEDVERBS_LABEL="Allowed Verbs"
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_webservices_content.ini:7]: alphasort please

File: administrator/language/en-GB/plg_webservices_modules.ini [hunk 11]
```diff
@@ -5,3 +5,6 @@
 
 PLG_WEBSERVICES_MODULES="Web Services - Modules"
 PLG_WEBSERVICES_MODULES_XML_DESCRIPTION="Add modules routes to the API for your website."
+PLG_WEBSERVICES_MODULES_ALLOWEDVERBS_LABEL="Allowed Verbs"
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_webservices_modules.ini:7]: alphasort please

File: administrator/language/en-GB/plg_webservices_users.ini [hunk 12]
```diff
@@ -5,3 +5,6 @@
 
 PLG_WEBSERVICES_USERS="Web Services - Users"
 PLG_WEBSERVICES_USERS_XML_DESCRIPTION="Add users routes to the API for your website."
+PLG_WEBSERVICES_USERS_ALLOWEDVERBS_LABEL="Allowed Verbs"
```
Comment (brianteeman) [inline administrator/language/en-GB/plg_webservices_users.ini:6]: alphasort please

File: libraries/src/Application/ApiApplication.php [hunk 13]
```diff
@@ -17,6 +17,7 @@
 use Joomla\CMS\Plugin\PluginHelper;
 use Joomla\CMS\Router\ApiRouter;
 use Joomla\CMS\Router\Exception\RouteNotFoundException;
+use Joomla\CMS\Router\Exception\RateLimitException;
```
Comment (brianteeman) [inline libraries/src/Application/ApiApplication.php:19]: alphasort please

File: plugins/webservices/contact/contact.php [hunk 14]
```diff
@@ -9,9 +9,11 @@
 
 defined('_JEXEC') or die;
 
+use Joomla\CMS\Factory;
 use Joomla\CMS\Plugin\CMSPlugin;
 use Joomla\CMS\Router\ApiRouter;
 use Joomla\Router\Route;
+use Joomla\Registry\Registry;
```
Comment (brianteeman) [inline plugins/webservices/contact/contact.php:15]: alphasort please

File: plugins/webservices/content/content.php [hunk 15]
```diff
@@ -3,14 +3,16 @@
  * @package     Joomla.Plugin
  * @subpackage  Webservices.Content
  *
- * @copyright   (C) 2019 Open Source Matters, Inc. <https://www.joomla.org>
+ * @copyright   Copyright (C) 2005 - 2020 Open Source Matters, Inc. All rights reserved.
```
Comment (brianteeman) [inline plugins/webservices/content/content.php]: please revert the change on this line

File: plugins/webservices/users/users.php [hunk 16]
```diff
@@ -9,7 +9,9 @@
 
 defined('_JEXEC') or die;
 
+use Joomla\CMS\Factory;
 use Joomla\CMS\Plugin\CMSPlugin;
+use Joomla\Registry\Registry;
```
Comment (brianteeman) [inline plugins/webservices/users/users.php]: alphasort please

File: administrator/language/en-GB/plg_webservices_content.ini (hunk 10 above)
Comment (alikon) [reply administrator/language/en-GB/plg_webservices_content.ini:7]: done

File: administrator/language/en-GB/plg_webservices_modules.ini (hunk 11 above)
Comment (alikon) [reply administrator/language/en-GB/plg_webservices_modules.ini:7]: done

File: administrator/language/en-GB/plg_webservices_users.ini (hunk 12 above)
Comment (alikon) [reply administrator/language/en-GB/plg_webservices_users.ini:6]: done

File: libraries/src/Application/ApiApplication.php (hunk 13 above)
Comment (alikon) [reply libraries/src/Application/ApiApplication.php:19]: done

File: plugins/webservices/contact/contact.php (hunk 14 above)
Comment (alikon) [reply plugins/webservices/contact/contact.php:15]: done

File: plugins/webservices/content/content.php (hunk 15 above)
Comment (alikon) [reply plugins/webservices/content/content.php]: done

File: plugins/webservices/users/users.php (hunk 16 above)
Comment (alikon) [reply plugins/webservices/users/users.php]: done

**Comment by ceford** [comment]
What does the rate limit mean - 60000 what. Is big good or bad? Does 0 mean no limit?<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/31579">issues.joomla.org/tracker/joomla-cms/31579</a>.</sub>

**Comment by alikon** [comment]
> Does 0 mean no limit?

yes

> What does the rate limit mean - 60000 what. Is big good or bad?

just a big number for testing

File: administrator/language/en-GB/com_plugins.ini [hunk 17]
```diff
@@ -33,6 +33,7 @@ COM_PLUGINS_NO_ITEM_SELECTED="No plugins selected."
 COM_PLUGINS_OPTION_ELEMENT="- Select Element -"
 COM_PLUGINS_OPTION_FOLDER="- Select Type -"
 COM_PLUGINS_PLUGIN="Plugin"
+COM_PLUGINS_RESET_LIMIT="Reset limit"
```
Comment (Quy) [inline administrator/language/en-GB/com_plugins.ini]:
```suggestion
COM_PLUGINS_RESET_LIMIT="Reset Limit"
```

File: api/components/com_content/src/Controller/ArticlesController.php [hunk 18]
```diff
@@ -11,10 +11,12 @@
 
 \defined('_JEXEC') or die;
 
+use Joomla\CMS\Factory;
 use Joomla\CMS\Filter\InputFilter;
 use Joomla\CMS\MVC\Controller\ApiController;
 use Joomla\Component\Fields\Administrator\Helper\FieldsHelper;
-
+use Joomla\CMS\Access\Exception\AuthenticationFailed;
+use Joomla\CMS\Router\Exception\RateLimitException;
```
Comment (Quy) [inline api/components/com_content/src/Controller/ArticlesController.php:19]: Alpha sort.

File: administrator/components/com_plugins/src/Controller/PluginController.php [hunk 19]
```diff
@@ -11,6 +11,8 @@
 \defined('_JEXEC') or die;
 
 use Joomla\CMS\MVC\Controller\FormController;
+use Joomla\CMS\Router\Route;
+use Joomla\CMS\Uri\Uri;
```
Comment (Quy) [inline administrator/components/com_plugins/src/Controller/PluginController.php]:
```suggestion
use Joomla\CMS\Router\Route;
```

File: api/components/com_content/src/Controller/ArticlesController.php (hunk 18 above)
Comment (alikon) [reply api/components/com_content/src/Controller/ArticlesController.php:19]: done

File: administrator/components/com_plugins/src/Controller/PluginController.php (hunk 19 above)
Comment (alikon) [reply administrator/components/com_plugins/src/Controller/PluginController.php]: done

**Comment by wilsonge** [comment]
OK I think rate limiting in core is excessive. This thing needs to be a plugin in itself and probably not part of core by default. We're supposed to provide a minimal implementation that can be extended I think.

The main thing I was looking for in the release blocker was the change to the ApiApplication (assuming that works on it's own)

**Comment by alikon** [comment]
These are an optional feature that is disabled by default. 
allowing public GET require a bare minimum rate control imho
